File: services/transcription/src/transcription/llm/chunking.py
"""Token-budget chunking for long transcripts (pure logic, zero package imports).

Counting defaults to a crude character heuristic (``len(text) // 2``);
callers that have a loaded model inject its real tokenizer via the
``count_tokens`` callables instead. The heuristic is deliberately
conservative -- Qwen-family BPE tokenizes Cyrillic at roughly 2-3
characters per token, and Russian is a primary language here -- because a
chunk that underfills the context window costs one extra LLM call, while a
chunk that overflows it costs the whole job.
"""

from __future__ import annotations

from collections.abc import Callable

TokenCounter = Callable[[str], int]

# Chat-template scaffolding, system prompt and per-call instructions that
# share the context window with the transcript chunk itself.
PROMPT_OVERHEAD_TOKENS = 512

# Below this the transcript would be sliced into confetti; budgets are
# floored here even under a pathologically small configured context.
MIN_BUDGET_TOKENS = 1024


def estimate_tokens(text: str) -> int:
    """A conservative token estimate: ~2 characters per token."""
    return max(1, len(text) // 2)


def input_budget_tokens(n_ctx: int, max_output_tokens: int, think_headroom_tokens: int) -> int:
    """How many input tokens a single completion may spend on transcript text.

    Everything else that shares the window -- the answer, the reasoning
    model's ``<think>`` block, prompt scaffolding -- is subtracted up front,
    so a chunk that fits this budget cannot overflow ``n_ctx``.
    """
    return max(
        MIN_BUDGET_TOKENS,
        n_ctx - max_output_tokens - think_headroom_tokens - PROMPT_OVERHEAD_TOKENS,
    )


def split_oversized(
    line: str, budget_tokens: int, count_tokens: TokenCounter = estimate_tokens
) -> list[str]:
    """Split a single line that alone exceeds ``budget_tokens``.

    Pieces break at whitespace where possible; a whitespace-free monster
    string is halved recursively. Nothing is ever dropped: the pieces
    concatenate (modulo the split whitespace) back to the original line.
    """
    if count_tokens(line) <= budget_tokens:
        return [line]

    words = line.split(" ")
    if len(words) == 1:
        mid = len(line) // 2
        if mid == 0:
            return [line]  # a single huge token; nothing left to split
        return split_oversized(line[:mid], budget_tokens, count_tokens) + split_oversized(
            line[mid:], budget_tokens, count_tokens
        )

    pieces: list[str] = []
    current: list[str] = []
    current_tokens = 0
    for word in words:
        word_tokens = count_tokens(word) + 1
        if current and current_tokens + word_tokens > budget_tokens:
            pieces.append(" ".join(current))
            current = []
            current_tokens = 0
        current.append(word)
        current_tokens += word_tokens
    if current:
        pieces.append(" ".join(current))
    # A single word can itself exceed the budget; recurse into those pieces.
    return [
        part for piece in pieces for part in split_oversized(piece, budget_tokens, count_tokens)
    ]


def chunk_line_ranges_with_overlap(
    lines: list[str],
    budget_tokens: int,
    overlap_tokens: int,
    count_tokens: TokenCounter = estimate_tokens,
) -> list[tuple[int, int]]:
    """Group ``lines`` into half-open ``(start, end)`` index ranges of at most
    ``budget_tokens``, with consecutive ranges sharing ~``overlap_tokens``
    worth of trailing lines.

    Index ranges rather than joined text, so the caller can map a chunk back
    to whatever its lines carry (segment timestamps, for the search index).
    Unlike :func:`chunk_lines` a single over-budget line is emitted as its
    own one-line range instead of being split -- splitting would break the
    line <-> range mapping, and the embedder truncates defensively anyway.
    """
    if budget_tokens <= 0:
        raise ValueError(f"budget_tokens must be positive, got {budget_tokens}")
    if overlap_tokens < 0 or overlap_tokens >= budget_tokens:
        raise ValueError(f"overlap_tokens must be in [0, budget_tokens), got {overlap_tokens}")

    ranges: list[tuple[int, int]] = []
    start = 0
    current_tokens = 0
    for index, line in enumerate(lines):
        line_tokens = count_tokens(line)
        if index > start and current_tokens + line_tokens > budget_tokens:
            ranges.append((start, index))
            # Back up over trailing lines worth ~overlap_tokens, but always
            # move forward past the previous range's start.
            overlap_start = index
            overlap_budget = overlap_tokens
            while overlap_start > start + 1:
                trailing = count_tokens(lines[overlap_start - 1])
                if trailing > overlap_budget:
                    break
                overlap_budget -= trailing
                overlap_start -= 1
            start = overlap_start
            current_tokens = sum(count_tokens(text) for text in lines[start:index])
        current_tokens += line_tokens
    if start < len(lines):
        ranges.append((start, len(lines)))
    return ranges


def chunk_lines(
    lines: list[str],
    budget_tokens: int,
    count_tokens: TokenCounter = estimate_tokens,
) -> list[str]:
    """Group ``lines`` into newline-joined chunks of at most ``budget_tokens``.

    Boundaries prefer falling between lines (transcript segments); a single
    line larger than the whole budget is split by ``split_oversized`` rather
    than emitted whole -- an over-budget chunk would overflow the context
    window, and dropping text would be worse.
    """
    if budget_tokens <= 0:
        raise ValueError(f"budget_tokens must be positive, got {budget_tokens}")

    fitted: list[str] = []
    for line in lines:
        if count_tokens(line) > budget_tokens:
            fitted.extend(split_oversized(line, budget_tokens, count_tokens))
        else:
            fitted.append(line)

    chunks: list[str] = []
    current: list[str] = []
    current_tokens = 0
    for line in fitted:
        line_tokens = count_tokens(line)
        if current and current_tokens + line_tokens > budget_tokens:
            chunks.append("\n".join(current))
            current = []
            current_tokens = 0
        current.append(line)
        current_tokens += line_tokens
    if current:
        chunks.append("\n".join(current))
    return chunks

File: crates/engine/src/llm/chunking.rs
//! Token-budget chunking for long transcripts.
//!
//! Port of `services/transcription/src/transcription/llm/chunking.py`.
//!
//! The budget is a crude character-count heuristic -- one token per three
//! characters -- deliberately biased low: a chunk that underfills the context
//! window costs one extra LLM call, while a chunk that overflows it costs the
//! whole job. Three characters is safe for Cyrillic as well as Latin text,
//! which BPE tokenizers split far more finely than English.

/// A budget no chunk could ever satisfy. Callers derive the budget from the
/// model's context window, so zero means a misconfigured model rather than a
/// long transcript.
#[derive(Debug, thiserror::Error)]
#[error("budget_tokens must be positive, got {0}")]
pub struct InvalidBudget(pub usize);

/// A conservative token estimate for `text`.
///
/// Counted in characters rather than bytes, matching Python's `len()` on a
/// `str`: a Cyrillic transcript is two bytes per character, and byte-counting
/// would halve every chunk for no reason.
pub fn estimate_tokens(text: &str) -> usize {
    (text.chars().count() / 3).max(1)
}

/// Group `lines` into newline-joined chunks of at most `budget_tokens`.
///
/// Boundaries always fall between lines (transcript segments), never inside
/// one. A single line larger than the whole budget still becomes its own
/// chunk -- the estimate is conservative enough that this only happens for
/// degenerate input, and dropping text would be worse.
pub fn chunk_lines<S: AsRef<str>>(
    lines: &[S],
    budget_tokens: usize,
) -> Result<Vec<String>, InvalidBudget> {
    if budget_tokens == 0 {
        return Err(InvalidBudget(budget_tokens));
    }

    let mut chunks: Vec<String> = Vec::new();
    let mut current: Vec<&str> = Vec::new();
    let mut current_tokens = 0;
    for line in lines {
        let line = line.as_ref();
        let line_tokens = estimate_tokens(line);
        if !current.is_empty() && current_tokens + line_tokens > budget_tokens {
            chunks.push(current.join("\n"));
            current.clear();
            current_tokens = 0;
        }
        current.push(line);
        current_tokens += line_tokens;
    }
    if !current.is_empty() {
        chunks.push(current.join("\n"));
    }
    Ok(chunks)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chunking_respects_the_budget_and_never_splits_a_line() {
        let body = "x".repeat(50);
        let lines: Vec<String> = (0..40).map(|i| format!("line {i} {body}")).collect();
        // ~100 tokens is ~300 characters, so about five of these lines.
        let budget = 100;
        let chunks = chunk_lines(&lines, budget).expect("a positive budget");

        assert!(chunks.len() > 1, "40 lines must not fit in one chunk");

        let regrouped: Vec<&str> = chunks.iter().flat_map(|chunk| chunk.lines()).collect();
        let original: Vec<&str> = lines.iter().map(String::as_str).collect();
        assert_eq!(regrouped, original, "every line survives, in order");

        // A chunk may exceed the budget only by the line that would not fit
        // before the boundary was taken -- never by more.
        for chunk in &chunks {
            assert!(estimate_tokens(chunk) <= budget + estimate_tokens(&lines[0]));
        }
    }

    #[test]
    fn a_single_oversized_line_still_becomes_its_own_chunk() {
        let huge = "y".repeat(10_000);
        let chunks = chunk_lines(std::slice::from_ref(&huge), 10).unwrap();
        assert_eq!(chunks, vec![huge]);
    }

    #[test]
    fn a_zero_budget_is_refused_rather_than_looping() {
        assert!(chunk_lines(&["a"], 0).is_err());
    }

    #[test]
    fn no_line_is_ever_estimated_at_zero_tokens() {
        // A run of empty segments must still advance the budget, or a
        // transcript of them would collapse into one unbounded chunk.
        assert_eq!(estimate_tokens(""), 1);
        assert_eq!(estimate_tokens("ab"), 1);
    }

    #[test]
    fn cyrillic_text_is_measured_in_characters_not_bytes() {
        // "да" is two characters but four bytes; byte-counting would double
        // the estimate and halve every chunk.
        assert_eq!(estimate_tokens(&"да".repeat(3)), estimate_tokens("abcdef"));
    }

    #[test]
    fn no_lines_at_all_produce_no_chunks() {
        let empty: [&str; 0] = [];
        assert!(chunk_lines(&empty, 100).unwrap().is_empty());
    }
}
